core/remotes/docker/lister.go

```diff
@@ -0,0 +1,144 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package docker
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/containerd/errdefs"
+	"github.com/containerd/log"
+)
+
+var ErrObjectNotRequired = errors.New("object not required")
+
+type TagList struct {
+	Name string   `json:"name"`
+	Tags []string `json:"tags"`
+}
+
+type dockerLister struct {
+	dockerBase *dockerBase
+}
+
+func (r *dockerLister) List(ctx context.Context) ([]string, error) {
+	refSpec := r.dockerBase.refspec
+	base := r.dockerBase
+	var (
+		errs  error
+		paths [][]string
+		caps  = HostCapabilityPull
+	)
+
+	// turns out, we have a valid digest, make an url.
+	paths = append(paths, []string{"tags/list"})
+	caps |= HostCapabilityResolve
+
+	hosts := base.filterHosts(caps)
+	if len(hosts) == 0 {
+		return nil, fmt.Errorf("no list hosts: %w", errdefs.ErrNotFound)
+	}
+
+	ctx, err := ContextWithRepositoryScope(ctx, refSpec, false)
+	if err != nil {
+		return nil, err
+	}
+
+	// right now, this is a single path only. But we can add more variations.
+	for _, u := range paths {
+		for _, host := range hosts {
+			ctxWithLogger := log.WithLogger(ctx, log.G(ctx).WithField("host", host.Host))
+
+			req := base.request(host, http.MethodGet, u...)
+			if err := req.addNamespace(base.refspec.Hostname()); err != nil {
+				return nil, err
+			}
+
+			req.header["Accept"] = []string{"application/json"}
+
+			log.G(ctxWithLogger).Debug("listing")
```

Comment:
This debug message doesn't seem helpful

Reply:
It adds a lot of context. Especially this line:

```go
ctxWithLogger := log.WithLogger(ctx, log.G(ctx).WithField("host", host.Host))
```

This is literally just saying listing, for the given host.